### PR TITLE
`Plagiarism checks`: Introduce min length parameter for programming exercises

### PR DESCRIPTION
**Pull request by jakubriegel**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-tests/).
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
_This change is not connected to CI/CD configuration._

### Motivation and Context
Artemis supports the minimum submission size parameter for plagiarism checks of text and modeling exercises but not for programming exercises. This parameter is very useful for removing false positives form the comparison.

### Description
Minimum size parameter for programming exercises plagiarism checks is introduced as minimal number of lines changed, according to `git diff`, between the template and the solution.

This PR also improves the UX by introducing minimum size tooltips dedicated to each exercise type.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise

1. Create two identical submissions that change the same file in the programming exercises template.
2. Run plagiarism detection on those submissions without minimum size set.
3. Verify that the results is present.
4. Set the minimum size to a number larger than the changed lines (ex. 50) and re-run detection.
5. Verify that your submissions are not included in the results or, if those were the only ones, that a warning was shown.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| PlagiarismDetectionService.java | 100% | ✅                           |
| plagiarism-inspector.component.ts | 88% | ✅                           |

### Screenshots
Minimum size option for programming exercises:
image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java [hunk 1]
```diff
@@ -158,6 +160,32 @@ public ProgrammingExerciseGitDiffReport getOrCreateReportOfExercise(ProgrammingE
...
+        var treeParserRepoA = new FileTreeIterator(repoA);
+        var treeParserRepoB = new FileTreeIterator(repoB);
+
+        try (var diffOutputStream = new ByteArrayOutputStream(); var git = Git.wrap(repoB)) {
+            git.diff().setOldTree(treeParserRepoB).setNewTree(treeParserRepoA).setOutputStream(diffOutputStream).call();
+            var diff = diffOutputStream.toString();
+            return extractDiffEntries(diff).stream().mapToInt(ProgrammingExerciseGitDiffEntry::getLineCount).map(Math::abs).sum();
+        }
+        catch (IOException | GitAPIException e) {
+            return Integer.MAX_VALUE;
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java:185]: Can you also log this error, so the admins can find potential future problems?

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/ProgrammingPlagiarismDetectionService.java [hunk 2]
```diff
@@ -343,41 +348,62 @@ public List<ProgrammingExerciseParticipation> filterStudentParticipationsForComp
...
+            return Optional.empty();
+        }
+    }
+
+    private boolean shouldAddRepo(ProgrammingExercise programmingExercise, int minimumSize, Repository repo, Optional<Repository> templateRepo) {
+        if (templateRepo.isEmpty()) {
+            return true;
+        }
+
+        var diffToTemplate = programmingExerciseGitDiffReportService.calculateNumberOfDiffLinesBetweenRepos(programmingExercise.getTemplateParticipation().getVcsRepositoryUrl(),
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/plagiarism/ProgrammingPlagiarismDetectionService.java:369]: This is shorter and does the same thing:
```suggestion
        var diffToTemplate = programmingExerciseGitDiffReportService.calculateNumberOfDiffLinesBetweenRepos(getTemplateRepositoryUrl(),
```

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/ProgrammingPlagiarismDetectionService.java [hunk 3]
```diff
@@ -343,41 +348,62 @@ public List<ProgrammingExerciseParticipation> filterStudentParticipationsForComp
...
+                repo.getLocalPath(), templateRepo.get().getRemoteRepositoryUrl(), templateRepo.get().getLocalPath());
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/plagiarism/ProgrammingPlagiarismDetectionService.java:370]: Are you using the template of the programming exercise in three of the four parameters?

File: src/main/webapp/i18n/de/plagiarism.json [hunk 4]
```diff
@@ -34,9 +34,9 @@
             "minimumScore": "Minimale Bewertung in %",
             "minimumScoreTooltip": "Berücksichtige nur Einreichungen mit einer Bewertung, die mindestens diesem Wert entspricht",
             "minimumSize": "Minimale Größe",
-            "minimumSizeTooltip": "Berücksichtige nur Einreichungen, deren Größe mindestens diesem Wert entspricht",
-            "minimumSizeTooltipModeling": "Berücksichtige nur Einreichungen, die mindestens so viele Modeling-Elemente wie der angegebene Wert haben",
-            "minimumSizeTooltipText": "Berücksichtige nur Einreichungen, die mindestens so viele Zeichen wie der angegebene Wert haben",
+            "minimumSizeTooltipTextExercise": "Consider only submissions that have at least as many different lines in comparison to the template as the specified value",
+            "minimumSizeTooltipProgrammingExercise": "Consider only submissions that have at least as many words as the specified value",
+            "minimumSizeTooltipModelingExercise": "Consider only submissions that have at least as many modeling elements as the specified value.",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/plagiarism.json:39]: You did not translate these into German

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java (hunk 1 above)
Comment (jakubriegel) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java:185]: Just got them from my translator @MarkusPaulsen :D

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/ProgrammingPlagiarismDetectionService.java (hunk 2 above)
Comment (jakubriegel) [reply src/main/java/de/tum/in/www1/artemis/service/plagiarism/ProgrammingPlagiarismDetectionService.java:369]: Thanks

File: src/main/webapp/i18n/de/plagiarism.json (hunk 4 above)
Comment (jakubriegel) [reply src/main/webapp/i18n/de/plagiarism.json:39]: Just got translations from my translator @MarkusPaulsen :D

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java (hunk 1 above)
Comment (jakubriegel) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java:185]: Good idea

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/ProgrammingPlagiarismDetectionService.java (hunk 3 above)
Comment (jakubriegel) [reply src/main/java/de/tum/in/www1/artemis/service/plagiarism/ProgrammingPlagiarismDetectionService.java]: That was an issue. Fixed now

**Review by JohannesStoehr — APPROVED** [review]
Code looks good to me now

**Review by tobias-lippert — APPROVED** [review]
Code LGTM

**Comment by jakubriegel** [comment]
To do:
- take abs from each change

**Review by julian-christl — APPROVED** [review]
Tested on TS1

**Review by rriyaldhi — APPROVED** [review]
Tested on TS1 during testing session.

**Comment by Kroko-fant** [comment]
looks good to me tested on ts1

**Review by florian-glombik — APPROVED** [review]
Tested on ts1 during testing session.

**Pull request merged**